### PR TITLE
Gas estimator

### PR DESCRIPTION
**Pull request by nachoiacovino**

### CODE REVIEW

File: components/gas-estimator/GasEstimatorBox.tsx [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    batchUpload,
+    mint,
+    claim,
+    distributeFunds,
+  }: GasPrice = GasEstimatorMap[contractType];
+
+  const [gasPrice, setGasPrice] = useState<number>(30);
+  const [ethPrice, setEthPrice] = useState<number>(3000);
+
+  useEffect(() => {
```
Comment (jnsdls) [inline components/gas-estimator/GasEstimatorBox.tsx:32]: this should be a useQuery

File: pages/api/gas.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+export default async (req: NextApiRequest, res: NextApiResponse) => {
+  if (req.method !== "GET") {
+    return res.status(400).json({ error: "invalid method" });
+  }
+
+  const gasEndpoint = `https://api.etherscan.io/api?module=gastracker&action=gasoracle&apikey=${process.env.ETHERSCAN_KEY}`;
+  const ethPriceEndpoint = `https://api.etherscan.io/api?module=stats&action=ethprice&apikey=${process.env.ETHERSCAN_KEY}`;
+
+  const gasPrice = await fetch(gasEndpoint, {
```
Comment (jnsdls) [inline pages/api/gas.ts:11]: if you want to make this slightly more efficient then make both these fetches at the same time with a `Promise.all([fetch1, fetch2])`

File: pages/api/gas.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  });
+
+  const ethPriceRes = await ethPrice.json();
+
+  const fullResponse = {
+    gas: gasPriceRes.result.SafeGasPrice || "30",
+    ethPrice: ethPriceRes.result.ethusd || "3000",
+  };
+
+  return res.status(200).json(fullResponse);
```
[the remark this answers is not in the document]
Comment (jnsdls) [reply pages/api/gas.ts:34]: ... and then throwing back the appropriate error code based on what error we get

**Review by jnsdls — CHANGES_REQUESTED** [review]

File: components/gas-estimator/GasEstimatorBox.tsx [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        ),
+      ).toFixed(4)} ETH`;
+    } else if (price && ethOrUsd === "usd") {
+      return `~$${(
+        Number(
+          ethers.utils.formatUnits(
+            `${((data?.gasPrice as number) || 30) * price}`,
+            "gwei",
+          ),
+        ) * (data?.ethPrice || 3400)
```
Comment (jnsdls) [inline components/gas-estimator/GasEstimatorBox.tsx:58]: I don't think we should hard-code a price here if we can't get it. if we can't get the price just disable the toggle I think

File: components/gas-estimator/GasEstimatorBox.tsx [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+      ).toFixed(2)}`;
+    }
+  };
+
+  return (
+    <Box p={6} border="1px solid" borderColor="gray.50" {...props}>
+      <Heading size="title.sm" mb={1}>
+        {CONTRACT_TYPE_NAME_MAP[contractType]}
+      </Heading>
+      <PriceLine gasPrice={deployContract as number}>
```
Comment (jnsdls) [inline components/gas-estimator/GasEstimatorBox.tsx:68]: all the `as` here scare me

File: pages/api/gas.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  const gasPrice = fetch(gasPriceEndpoint);
+  const ethPrice = fetch(ethPriceEndpoint);
+
+  try {
+    const [gasPriceRes, ethPriceRes] = await Promise.all([gasPrice, ethPrice]);
+    const gasData = await gasPriceRes.json();
+    const ethPriceData = await ethPriceRes.json();
+
+    if (gasData.status !== "1") {
+      return res.status(500).json({ error: "Failed to fetch gas price" });
```
Comment (jnsdls) [inline pages/api/gas.ts:20]: this should not be a status code , if we get here it means the responses didn't fail so there isn't a server error it's just a data error in some way, so I'd say maybe 400

File: pages/api/gas.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+      return res.status(500).json({ error: "Failed to fetch ETH price" });
+    }
+
+    return res.status(200).json({
+      gasPrice: gasData.result.ProposeGasPrice,
+      ethPrice: ethPriceData.result.ethusd,
+    });
+  } catch (err) {
+    console.error(err);
+    return res.status(500).json({ error: "Internal server error" });
```
Comment (jnsdls) [inline pages/api/gas.ts:33]: this is likely a 502 or 504 (the requests to the external service failed)

File: components/gas-estimator/GasEstimatorBox.tsx [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+import { ContractType } from "@thirdweb-dev/sdk";
+import {
+  CONTRACT_TYPE_NAME_MAP,
+  GasEstimatorMap,
+  GasPrice,
+} from "constants/mappings";
+import { ethers } from "ethers";
+
+interface PriceLineProps {
+  gasPrice: number;
```
Comment (jnsdls) [inline components/gas-estimator/GasEstimatorBox.tsx:11]:
```suggestion
  gasPrice?: number;
```

File: components/gas-estimator/GasEstimatorBox.tsx (hunk 5 above)
Comment (jnsdls) [reply components/gas-estimator/GasEstimatorBox.tsx:68]: if u implement the`?` on the PriceLineProps then u don't have to do all the `as` anymore

**Review by jnsdls — CHANGES_REQUESTED** [review]

File: @3rdweb-sdk/react/cache-keys.ts [hunk 9]
```diff
@@ -25,6 +25,7 @@ export const dashboardKeys = {
   all: ["dashboard"] as const,
   lists: () => [...dashboardKeys.all, "list"] as const,
   list: (address = AddressZero) => [...dashboardKeys.lists(), address] as const,
+  gas: () => [...dashboardKeys.all, "gas"] as const,
```
Comment (jnsdls) [inline @3rdweb-sdk/react/cache-keys.ts:28]: don't need to add this here

File: @3rdweb-sdk/react/hooks/useGas.ts [hunk 10]
```diff
@@ -0,0 +1,15 @@
+import { useQuery } from "react-query";
+import { dashboardKeys } from "..";
+
+export function useGas() {
+  return useQuery(
+    dashboardKeys.gas(),
```
Comment (jnsdls) [inline @3rdweb-sdk/react/hooks/useGas.ts:6]: just make this an array

```suggestion
    ["gas-price", "ethereum"]
```

**Review by jnsdls — CHANGES_REQUESTED** [review]
small useQuery changes

File: components/gas-estimator/GasEstimatorBox.tsx (hunk 4 above)
Comment (nachoiacovino) [reply components/gas-estimator/GasEstimatorBox.tsx:58]: what about if we can't get the gasPrice?

File: pages/api/gas.ts (hunk 6 above)
Comment (nachoiacovino) [reply pages/api/gas.ts:20]: I just changed this, tell me if this is what you had in mind

**Pull request merged**